pulsar/consumer_multitopic.go

```diff
@@ -189,25 +189,30 @@ func ackIDListFromMultiTopics(log log.Logger, msgIDs []MessageID, findConsumer f
 		}
 	}
 
-	ackError := AckError{}
-	for consumer, ids := range consumerToMsgIDs {
-		if err := consumer.AckIDList(ids); err != nil {
-			if topicAckError := err.(AckError); topicAckError != nil {
+	subErrCh := make(chan error, len(consumerToMsgIDs))
+	errCh := make(chan error)
+	go func() {
+		ackError := AckError{}
+		for i := 0; i < len(consumerToMsgIDs); i++ {
+			err := <-subErrCh
+			if topicAckError, ok := err.(AckError); ok {
 				for id, err := range topicAckError {
 					ackError[id] = err
 				}
-			} else {
-				// It should not reach here
-				for _, id := range ids {
-					ackError[id] = err
-				}
 			}
 		}
+		if len(ackError) == 0 {
+			errCh <- nil
+		} else {
+			errCh <- ackError
+		}
+	}()
+	for consumer, ids := range consumerToMsgIDs {
+		go func() {
+			subErrCh <- consumer.AckIDList(ids)
```

Comment:
```go
	for consumer, ids := range consumerToMsgIDs {
		consumer := consumer
		ids := ids
		go func() {
			subErrCh <- consumer.AckIDList(ids)
		}()
	}
```
Necessary to make a variable copy? Although this issue was fixed in [1.22](https://go.dev/blog/loopvar-preview), would it trigger this issue if the user is using a version less than 1.22?

https://github.com/apache/pulsar-client-go/blob/35076ac09138b5fa11a508ccc0ef8cd1b1993347/pulsar/consumer_impl.go#L779

Reply:
It's already 1.22: https://github.com/apache/pulsar-client-go/blob/92c6e284cfcd1eb921116da6482b79e5500e03ae/go.mod#L3